activerecord/lib/active_record/connection_adapters/abstract/transaction.rb

```diff
@@ -188,6 +188,31 @@ def full_rollback?; true; end
       def joinable?; @joinable; end
       def closed?; false; end
       def open?; !closed?; end
+
+      private
+        def prepare_instances_to_run_callbacks_on(records)
```

Comment:
We are now looping through records twice and building a hash even if we don't need to run commit callbacks. Can we merge both loops somehow?

Reply:
When `@run_commit_callbacks` is true, this change is net zero on the number of hashes, since I was able to replace `already_run_callbacks` from inside of `commit_records`. It is an additional linear scan, but I don't think it's avoidable. For any given instance, in order to know whether to call `committed!` with a true or false value of `should_run_callbacks`, we need to have seen all other instances of the same record.

I will refactor to avoid the extra hash/iteration when `@run_commit_callbacks` is false.